### PR TITLE
fix: IdentityCreateStateTransition signature

### PR DESCRIPTION
**Pull request by shuplenkov**

### CODE REVIEW

File: lib/errors/IdentityLockTransactionNotFound.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+const ConsensusError = require('./ConsensusError');
+
+class IdentityLockTransactionNotFound extends ConsensusError {
```
Comment (shumkov) [inline lib/errors/IdentityLockTransactionNotFound.js]:
```suggestion
class IdentityLockTransactionNotFoundError extends ConsensusError {
```

File: lib/errors/IdentityLockTransactionNotFound.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+   * Get transaction hash
+   *
+   * @return {string}
+   */
+  getTransactionHash() {
+    return this.transactionHash;
+  }
+}
+
+module.exports = IdentityLockTransactionNotFound;
```
Comment (shumkov) [inline lib/errors/IdentityLockTransactionNotFound.js]:
```suggestion
module.exports = IdentityLockTransactionNotFoundError;
```

File: lib/errors/InvalidIdentityLockTransaction.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+const ConsensusError = require('./ConsensusError');
+
+class InvalidIdentityLockTransaction extends ConsensusError {
```
Comment (shumkov) [inline lib/errors/InvalidIdentityLockTransaction.js]:
```suggestion
class InvalidIdentityLockTransactionError extends ConsensusError {
```

File: lib/errors/InvalidIdentityLockTransaction.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+   * Get lock transaction
+   *
+   * @return {Transaction}
+   */
+  getTransaction() {
+    return this.transaction;
+  }
+}
+
+module.exports = InvalidIdentityLockTransaction;
```
Comment (shumkov) [inline lib/errors/InvalidIdentityLockTransaction.js]:
```suggestion
module.exports = InvalidIdentityLockTransactionError;
```

File: lib/identity/stateTransitions/AbstractIdentityStateTransition.js [hunk 5]
```diff
@@ -4,7 +4,40 @@ const AbstractStateTransition = require('../../stateTransition/AbstractStateTran
  * @abstract
  */
 class AbstractIdentityStateTransition extends AbstractStateTransition {
```
Comment (shumkov) [inline lib/identity/stateTransitions/AbstractIdentityStateTransition.js]: Why do we need the abstract class for Identity STs?  Other Identity STs will use Identity's public key as other STs.

Comment (shumkov) [reply lib/identity/stateTransitions/AbstractIdentityStateTransition.js]: Oh.. My bad. TopUp will work in the same way.

File: lib/stateTransition/AbstractStateTransitionSigned.js [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+    return this;
+  }
+
+  /**
+   * Sign data with private key
+   * @param {string|Buffer|Uint8Array|PrivateKey} privateKey string must be hex or base58
+   * @return {AbstractStateTransition}
+   */
+  signByPrivateKey(privateKey) {
```
Comment (shumkov) [inline lib/stateTransition/AbstractStateTransitionSigned.js]: Looks like duplicate code

File: lib/stateTransition/AbstractStateTransitionSigned.js [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+    return isSignatureVerified;
+  }
+
+  /**
+   * Verify signature with public key
+   * @param {string|Buffer|Uint8Array|PublicKey} publicKey string must be hex or base58
+   * @returns {boolean}
+   */
+  verifySignatureByPublicKey(publicKey) {
```
Comment (shumkov) [inline lib/stateTransition/AbstractStateTransitionSigned.js]: Looks like duplicate code

File: lib/stateTransition/AbstractStateTransitionSigned.js [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    const signature = this.getSignature();
+    if (!signature) {
+      throw new StateTransitionIsNotSignedError(this);
+    }
+
+    if (this.getSignaturePublicKeyId() !== publicKey.getId()) {
+      throw new PublicKeyMismatchError(publicKey);
+    }
+
+    const signatureBuffer = Buffer.from(signature, 'base64');
```
Comment (shumkov) [inline lib/stateTransition/AbstractStateTransitionSigned.js]: Can we use `verifySignatureByPublicKey` inside?

File: lib/identity/stateTransitions/identityCreateTransition/validateLockTransaction.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+ * @typedef ValidateLockTransaction
+ * @param {Transaction} transaction
+ * @param {number} outputIndex
+ * @returns {ValidationResult}
+ */
+function validateLockTransaction(transaction, outputIndex) {
+  const result = new ValidationResult();
+
+  if (!transaction.outputs[outputIndex]) {
+    result.addError(new InvalidIdentityLockTransactionError(transaction));
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityCreateTransition/validateLockTransaction.js]: Let's pass messages there

File: lib/identity/stateTransitions/identityCreateTransition/validateLockTransaction.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+  if (!result.isValid()) {
+    return result;
+  }
+
+  const { script } = transaction.outputs[outputIndex];
+  const publicKeyHash = script.getData();
+
+  if (!script.isDataOut() || publicKeyHash.length !== 20) {
+    result.addError(new InvalidIdentityLockTransactionError(transaction));
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityCreateTransition/validateLockTransaction.js]: The same

File: lib/errors/InvalidIdentityLockTransactionError.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
+const ConsensusError = require('./ConsensusError');
+
+class InvalidIdentityLockTransactionError extends ConsensusError {
+  /**
+   * @param {Transaction} transaction
+   */
+  constructor(transaction) {
+    super('Invalid identity lock transaction');
```
Comment (shumkov) [inline lib/errors/InvalidIdentityLockTransactionError.js]: `Invalid identity lock transaction: ${message}`

File: lib/identity/stateTransitions/identityCreateTransition/validateIdentityCreateSTDataFactory.js [hunk 12]
```diff
@@ -32,7 +37,45 @@ function validateIdentityCreateSTDataFactory(dataProvider) {
...
+      result.addError(new IdentityLockTransactionNotFoundError(transactionHash));
+    }
+
+    if (!result.isValid()) {
+      return result;
+    }
+
+    const transaction = new Transaction(rawTransaction);
+
+    result.merge(
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityCreateTransition/validateIdentityCreateSTDataFactory.js]: Extract `validateLockTransaction` logic here

File: lib/identity/stateTransitions/identityCreateTransition/validateIdentityCreateSTDataFactory.js [hunk 13]
```diff
@@ -32,7 +37,45 @@ function validateIdentityCreateSTDataFactory(dataProvider) {
...
     }
 
-    // TODO: Here we need to fetch lock transaction, extract pubkey from it and verify signature
+    if (!result.isValid()) {
+      return result;
+    }
+
+    // fetch lock transaction, extract pubkey from it and verify signature
+
+    const lockedOutBuffer = Buffer.from(identityCreateTransition.getLockedOutPoint(), 'base64');
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityCreateTransition/validateIdentityCreateSTDataFactory.js]: This is a part of `validateLockTransaction` let's move it there too. Signature validation as well

File: lib/identity/stateTransitions/identityCreateTransition/validateIdentityCreateSTDataFactory.js [hunk 14]
```diff
@@ -32,7 +37,45 @@ function validateIdentityCreateSTDataFactory(dataProvider) {
...
+    const { transactionHash, outputIndex } = Transaction.parseOutPointBuffer(lockedOutBuffer);
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityCreateTransition/validateIdentityCreateSTDataFactory.js]: We need to catch parsing errors and add consensus error if the outPoint Buffer format is wrong

File: lib/stateTransition/AbstractStateTransition.js [hunk 15]
```diff
@@ -37,10 +35,30 @@ class AbstractStateTransition {
...
+  getSignature() {
+    return this.signature;
+  }
+
+  /**
+   * Set signature
+   * @param {string|null} [signature]
+   * @return {AbstractStateTransition}
+   */
+  setSignature(signature = null) {
```
Comment (shumkov) [inline lib/stateTransition/AbstractStateTransition.js]: Let's make the argument is required

File: lib/identity/stateTransitions/identityCreateTransition/validateLockTransactionFactory.js [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+
+    if (!result.isValid()) {
+      return result;
+    }
+
+    const transaction = new Transaction(rawTransaction);
+
+    //
+    //
+    // transaction.outputs[0].script.chunks[1] = undefined;
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityCreateTransition/validateLockTransactionFactory.js]: Please remove commented code

File: lib/identity/stateTransitions/identityCreateTransition/validateLockTransactionFactory.js [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+    if (!transaction.outputs[outputIndex]) {
+      result.addError(new InvalidIdentityLockTransactionError(`Output with index ${outputIndex} not found`, transaction));
```
Comment (shumkov) [inline lib/identity/stateTransitions/identityCreateTransition/validateLockTransactionFactory.js]: This situation looks more like ”Invalid identity outpoint”

File: lib/DataProviderInterface.js [hunk 18]
```diff
@@ -35,7 +35,7 @@
  * @method
  * @name DataProvider#fetchTransaction
  * @param {string} id
- * @returns {Promise<{ confirmations: number }|null>}
+ * @returns {Promise<RawTransaction|null>}
```
Comment (shumkov) [inline lib/DataProviderInterface.js]: Where did you define it?

File: lib/stateTransition/createStateTransitionFactory.js [hunk 19]
```diff
@@ -52,9 +52,11 @@ function createStateTransitionFactory() {
         throw new InvalidStateTransitionTypeError(rawStateTransition);
     }
 
-    stateTransition
-      .setSignature(rawStateTransition.signature)
-      .setSignaturePublicKeyId(rawStateTransition.signaturePublicKeyId);
+    stateTransition.setSignature(rawStateTransition.signature);
```
Comment (shumkov) [inline lib/stateTransition/createStateTransitionFactory.js]: Maybe we should set signatures via constructors (abstract) as a part of a raw transition data?

File: lib/DataProviderInterface.js (hunk 18 above)
Comment (shuplenkov) [reply lib/DataProviderInterface.js]: it's undefined yet. Should I use Object instead?

Comment (shumkov) [reply lib/DataProviderInterface.js]: Yeah, it's better than specify undefined type.

File: lib/stateTransition/AbstractStateTransition.js [hunk 20]
```diff
@@ -4,21 +4,33 @@ const {
...
-const StateTransitionIsNotSignedError = require('./errors/StateTransitionIsNotSignedError');
-const PublicKeyMismatchError = require('./errors/PublicKeyMismatchError');
 
 /**
  * @abstract
  */
 class AbstractStateTransition {
-  constructor() {
-    this.signaturePublicKeyId = null;
+  /**
```
Comment (shumkov) [inline lib/stateTransition/AbstractStateTransition.js]:
```suggestion
  /**
     * @param {
    * RawDataContractStateTransition|
    * RawDocumentsStateTransition|
    * RawIdentityCreateTransition
   * } [rawStateTransition]
   */
  constructor(rawStateTransition = {}) {
    this.signature = rawStateTransition.signature;
  }
```

File: lib/stateTransition/AbstractStateTransitionIdentitySigned.js [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+ */
+class AbstractStateTransitionIdentitySigned extends AbstractStateTransition {
+  /**
+   * @param {
+   * RawDataContractStateTransition|
+   * RawDocumentsStateTransition|
+   * null
+   * } rawStateTransition
+   */
+  constructor(rawStateTransition = null) {
```
Comment (shumkov) [inline lib/stateTransition/AbstractStateTransitionIdentitySigned.js]: the same thing

File: lib/stateTransition/createStateTransitionFactory.js [hunk 22]
```diff
@@ -52,9 +52,11 @@ function createStateTransitionFactory() {
         throw new InvalidStateTransitionTypeError(rawStateTransition);
     }
 
-    stateTransition
-      .setSignature(rawStateTransition.signature)
-      .setSignaturePublicKeyId(rawStateTransition.signaturePublicKeyId);
+    if (stateTransition.getType() !== types.IDENTITY_CREATE) {
```
Comment (shumkov) [inline lib/stateTransition/createStateTransitionFactory.js]: Please move it to the cases above.

**Pull request merged**